Fix reshape(int, ...)

sklam opened the pull request:
Fix for #1506 
* fixes the inability for reshape to accept scalar arguments
* make some related tests to be more precise in the feature under test.

sklam commented:
Rerunning the tests today is causing an error on my laptop.
Investigating...

pitrou commented on numba/targets/arrayobj.py:
```diff
@@ -1115,6 +1117,26 @@ def array_reshape(context, builder, sig, args):
+    # values
+    ary = args[0]
+    dims = args[1:]
+    # coerce all types to uintp
```
I think it should cast to intp (-1 is a documented valid value although our implementation doesn't support it).

pitrou commented on numba/tests/test_array_manipulation.py:
```diff
@@ -52,146 +73,210 @@ def bad_float_index(arr):
 
+        expected = pyfunc(a)
+        got = cfunc(a)
+        np.testing.assert_equal(expected, got)
```
I don't think assert_equal() is enough here, it will only test the values. We should use assertPreciseEqual() to also test dtype, shape, etc.

pitrou commented on numba/typing/arraydecl.py:
```diff
@@ -257,19 +253,36 @@ def resolve_nonzero(self, ary, args, kws):
+            # single arg
+            shape, = args
+
+            if shape in types.number_domain:
```
Hmm, I don't think there's a reason to accept floats and complex numbers, so let's check for types.Integer instead perhaps?

pitrou commented on numba/typing/arraydecl.py:
```diff
@@ -257,19 +253,36 @@ def resolve_nonzero(self, ary, args, kws):
-        retty = ary.copy(ndim=ndim)
-        return signature(retty, shape)
+            # vararg case
+            if any(a not in types.number_domain for a in args):
```
Same here...

sklam commented:
Don't merge this yet.
I am fixing something.

pitrou commented on numba/targets/arrayobj.py:
```diff
@@ -1101,6 +1107,8 @@ def array_reshape(context, builder, sig, args):
 
     with builder.if_then(fail):
+        # Release allocated array if the function error out
+        context.nrt_decref(builder, aryty, args[0])
```
I don't understand this. This is not a newly-allocated array, just one of the functions' arguments...

sklam replied:
I now realize it is a bad fix for the problem that I didn't think through.  The actual problem is that cleanup code is not executed when an exception is raised.  As a result, the incref for the argument is not reversed.  I need to do a proper fix.

sklam commented:
I reverted the adhoc cleanup code (https://github.com/numba/numba/pull/1508#discussion_r45867284) inside reshape implementation. I added an expectedFailure test and opened #1549 to track it.

pitrou commented on numba/tests/test_array_manipulation.py:
```diff
@@ -52,146 +73,223 @@ def bad_float_index(arr):
+            self.test_reshape_array_to_1d(flags=no_pyobj_flags, layout='A')
+        self.assertIn("reshape() supports contiguous array only",
+                      str(raises.exception))
+        self.disable_leak_check()
```
Perhaps comment why it is disabled?

pitrou commented on numba/tests/test_array_manipulation.py:
```diff
@@ -52,146 +73,223 @@ def bad_float_index(arr):
+        with self.assertRaises(NotImplementedError) as raises:
+            self.test_reshape_array_to_1d(flags=no_pyobj_flags, layout='F')
+        self.assertIn("incompatible shape for array", str(raises.exception))
+        self.disable_leak_check()
```
Ditto.

pitrou commented on numba/tests/test_array_methods.py:
```diff
@@ -198,6 +198,12 @@ def check_err_size(arr, shape):
             self.assertEqual(str(raises.exception),
                              "total size of new array must be unchanged")
 
+        def check_err_negative(arr, shape):
```
I'm only noticing now, but it seems confusing to have tests for reshape() in two different files (test_array_methods and test_array_manipulation). Perhaps we should consolidate them?

seibert commented:
I think this is ready to merge once Antoine's suggestions are addressed.

sklam commented:
PRs #1542 and #1509 are modifying the test files.  I will do the consolidation of the tests in another PR.  I opened an issue to track it: #1562 .

seibert commented:
OK, I'll merge this PR, and then we can move the review along the PR chain.
